### ISSUE TITLE
Session object not modifiable in typescript (while using @types/express-session)

### ISSUE THREAD

**Issue by qoobes** [expressjs/session#799]
I've been using express-session in various express applications, but never with typescript. 
**Disclaimer:** This might be a bug with @types/express-session, but I figure this is the right community to ask either way, so here's what's happening.

The session object is essentially just not modifiable, ie I can't seem to add properties to it, and, perhaps because of a recent update, all methods of fixing this I've found either don't work or just break the whole thing. 

In this example use case:

```ts
router.get("/example", (req, res) => {
	console.log(req.session); // Works
	req.session.username = "mine" // Fails
	req.session.destroy(data => console.log(data)); // Works
	res.send('something')
});
```

I get the following error: 

`Property 'username' does not exist on type 'Session & Partial<SessionData>'`

Not sure if this is in any way useful, but these are my session options 

```ts
const sessionConfig: session.SessionOptions = {
	name: "uid",
	secret,
	store: new RedisStore({
		client: redisClient,
		disableTTL: true,
		disableTouch: true,
	}),
	cookie: {
		maxAge: 1000000000000999,
		httpOnly: true,
		sameSite: "lax",
		secure: __prod__,
	},
	saveUninitialized: false,
	resave: false,
};
```

I have attempted to create a separate type combining express.Request and the session options, but I can't really get that to work either, and I'm assuming there's a bit more elegant way of dealing with this.

Thank your for your time.

**Comment by randomicons** [comment]
Add this to a file in your project ( ie types.ts etc.)
```typescript
declare module "express-session" {
  interface SessionData {
    username: string
  }
}
```
You can add all of your custom fields in session inside of the `SessionData` interface.

**Comment by wcauchois** [comment]
@randomicons that approach isn't working for me in TypeScript 4.1. I can't even tell what exactly the issue is, I think I've configured my typeRoots etc correctly. Is anyone else having issues with that approach?

**Comment by jasonlimantoro** [comment]
1. Make sure your `tsconfig.json`'s typeroots declares custom types first, before the node_modules's.
```json
"typeRoots": ["./src/typings", "./node_modules/@types"]
```
2. Declare definition file
```typescript
// ./src/typings/express-session/index.d.ts
import "express-session"; // don't forget to import the original module

declare module "express-session" {
    interface SessionData {
        userId: number
    }
}
```

That should be enough, hope this helps everyone who is stuck on Typescript hole 😄

**Comment by ktjd123** [comment]
I'm having trouble with this too.

with @jasonlimantoro answer It gives me this error

```ts

server/index.ts(55,35): error TS2345: Argument of type 'typeof import("/Users/peter/Server/auto-posting/@types/express-session/index")' is not assignable to parameter of type '(options?: any) => RequestHandler<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
  Type 'typeof import("/Users/peter/Server/auto-posting/@types/express-session/index")' provides no match for the signature '(options?: any): RequestHandler<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/index.ts(57,5): error TS2349: This expression is not callable.
  Type 'typeof import("/Users/peter/Server/auto-posting/@types/express-session/index")' has no call signatures.
server/routes/auth.ts(9,12): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/auth.ts(9,28): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/auth.ts(10,46): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/auth.ts(13,9): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/auth.ts(17,7): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/auth.ts(44,7): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/auth.ts(76,7): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/site.ts(8,12): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/site.ts(8,28): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
server/routes/site.ts(9,11): error TS2339: Property 'session' does not exist on type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
```
But when I make a copy of @types/express-session and paste it to the .d.ts file and edit SessionData it works file.

It might be my tsconfig's problem.

my tsconfig is

```json
{
  "compilerOptions": {
    "outDir": "./build",
    "allowJs": true,
    "target": "esnext",
    "lib": ["es2015", "es2016", "es2017", "es2018", "dom"],
    "module": "commonjs",
    "moduleResolution": "node",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "downlevelIteration": true,
    "sourceMap": true,
    "typeRoots": ["./@types", "./node_modules/@types"],
    "baseUrl": "./",
    "alwaysStrict": true,
    "paths": {
      "*": ["@types/*"]
    },
    "strict": true,
    "removeComments": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "jsx": "preserve",
    "isolatedModules": true,
    "noEmit": false,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true
  },
  "include": ["./server/**/*"],
  "exclude": ["./node_modules", "./build", "./.next"]
}

```

**Comment by mstaicu** [comment]
I had the same issue, though I'm counting the days until there will be a regression and the declaration merging will stop working. 

tl;dr; importing the module in the declaration file, for which I wanted to perform the declaration merge, fixed my issue

```
import 'express-session';
```

Folder structure:

```
src
  types
    express
      index.d.ts
    express-session
      index.d.ts
```

`index.d.ts` of `src/types/express`

```
declare namespace Express {
  interface Request {
    user: import('../api').User;
  }
}
```

`index.d.ts` of `src/types/express-session`

```
import 'express-session';

declare module 'express-session' {
  interface SessionData {
    user: import('../api').User;
  }
}
```

`package.json`

```
{
  "dependencies": {
    "express": "~4.16.1",
    "express-session": "^1.16.2"
  },
  "devDependencies": {
    "@types/express": "^4.17.6",
    "@types/express-session": "^1.17.3",
    "typescript": "^3.8.3"
  }
}
```

`tsconfig.json`

```
{
  "compilerOptions": {
    "alwaysStrict": true,
    "esModuleInterop": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noFallthroughCasesInSwitch": true,
    "noEmitOnError": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "moduleResolution": "node",
    "lib": ["ES2019"],
    "module": "commonjs",
    "target": "ES2019",
    "outDir": "dist",
    "rootDir": "src"
  },
  "include": ["src"]
}
```

**Comment by danieldai** [comment]
@jasonlimantoro Thanks. It works for me

**Comment by rsimp** [comment]
Would like to point out that `import 'express-session';` is solving the issue, but probably not how you think. In an ambient declaration file if the file is *not* a module, declarations will overwrite the entire module, essentially zeroing out everything not explicitly added. On the other hand, if the file is considered a module then you will be "augmenting" the declaration.

You could simply add `export {}` and it would have the same effect. Adding the import statement is simply turning the file into a module. It's converting the way typescript handles module declarations i.e. overwriting vs augmentation.

**Comment by raarts** [comment]
How I solved it:

```
interface SessionData extends Session {
  username: string;
}

router.get("/example", (req: { session: SessionData } , res: any) => {
  console.log(req.session.username);  // username was initialized elsewhere
});
```

**Comment by dougwilson** [comment]
Thanks everyone! I'm going to close this as it seems the conversation has settled and there isn't any action for this module.

**Comment by florianwalther-private** [comment]
@rsimp 
Thank you for shining some light on this!

> If you augment inside of a normal module, instead of a declaration, there's no need to do any extra imports.
Do you have an example for this because I don't know what it is.

**Comment by AwesomeZaidi** [comment]
> How I solved it:
> 
> ```
> interface SessionData extends Session {
>   username: string;
> }
> 
> router.get("/example", (req: { session: SessionData } , res: any) => {
>   console.log(req.session.username);  // username was initialized elsewhere
> });
> ```

Typing the request like this doesn't seem like a good idea because then you can't access the other properties on the request. Should we create a custom defined type for the express Request to include out custom express-session SessionData?

**Comment by ContentGamer** [comment]
This was also a problem for me, this is what I have tried in typescript and it works.

```ts
   console.log((req.session as any).username)
```

**Comment by AlexPalaz** [comment]
You can try with this. It works for me:

type CustomSessionData = session.SessionData & {
  accessToken: string | null;
  refreshToken: string | null;
};

type CustomSession = Session & Partial<CustomSessionData>;

accessToken and refreshToken are custom properties I added.
